src/utils/loadable.ts

```diff
@@ -0,0 +1,67 @@
+import { atom } from 'jotai'
+import type { Atom } from 'jotai'
+import { getWeakCacheItem, setWeakCacheItem } from './weakCache'
+
+const loadableAtomCache = new WeakMap()
+
+export type Loadable<Value> =
+  | LoadableLoading
+  | LoadableError
+  | LoadableValue<Value>
+
+export interface LoadableError {
+  state: 'hasError'
+  error: unknown
+}
+
+export interface LoadableLoading {
+  state: 'loading'
+}
+
+export interface LoadableValue<Value> {
+  state: 'hasData'
+  data: Value
+}
+
+const LOADING_LOADABLE: LoadableLoading = { state: 'loading' }
+
+export function loadable<Value>(anAtom: Atom<Value>): Atom<Loadable<Value>> {
+  const deps: object[] = [anAtom]
+  const cachedAtom = getWeakCacheItem(loadableAtomCache, deps)
+  if (cachedAtom) {
+    return cachedAtom as Atom<Loadable<Value>>
+  }
+
+  const refAtom = atom<{ error: LoadableError | null }>({
+    error: null,
+  })
```

Comment:
refAtom must be a read-only atom with a function for lazy initialization, otherwise the state could be shared among different providers.
I guess we can just store the entire `Loadable<Value>` type in the ref.
```suggestion
  const loadableRefAtom = atom((): [Loadable<Value>] => ([{ state: 'loading' }]))
```

Hmm, the tuple might be not very readable. In such a case, we can make it an object.

Reply:
Let me try writing the new derivedAtom with this ref.

```js
  const derivedAtom = atom((get): Loadable<Value> => {
    const loadableRef = get(loadableRefAtom)
    try {
      const value = get(anAtom)
      loadableRef[0] = { state: 'hasData', data: value }
    } catch (error) {
      if (error instanceof Promise) {
        if (loadableRef[0].state !== 'loading') {
          loadableRef[0] = { state: 'loading' }
        }
      } else {
        if (loadableRef[0].state !== 'hasError' || !Object.is(loadableRef[0].error, error)) {
          loadableRef[0] = { state: 'hasError', error }
        }
      }
    }
    return loadableRef[0]
  })
```

Now, I'm not sure if this is more readable, maybe your version is better.

---
Now, I have another idea. We can create a WeakMap at module level.
The key of the WeakMap is `error` object and the value is Loadable<Value> (= `{ state: 'hasError', error }`)
This allows us to eliminate the hacky refAtom.
Yeah, I think is far cleaner. What do you think?

Reply:
I like it!